components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -2,89 +2,2001 @@
+                underlying_source_type,
+                strategy_hwm,
+                global,
+                can_gc,
+            )),
+            global,
+            can_gc,
+        )
+    }
+
     pub(crate) fn set_stream(&self, stream: &ReadableStream) {
         self.stream.set(Some(stream))
     }
 
     /// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-pull-into>
+    #[allow(unsafe_code)]
     pub(crate) fn perform_pull_into(
         &self,
-        _read_into_request: &ReadIntoRequest,
-        _view: HeapBufferSource<ArrayBufferViewU8>,
-        _options: &ReadableStreamBYOBReaderReadOptions,
-        _can_gc: CanGc,
+        read_into_request: &ReadIntoRequest,
+        view: HeapBufferSource<ArrayBufferViewU8>,
+        options: &ReadableStreamBYOBReaderReadOptions,
+        can_gc: CanGc,
     ) {
-        todo!()
+        let cx = GlobalScope::get_cx();
+        // Let stream be controller.[[stream]].
+        let stream = self.stream.get().unwrap();
+
+        // Let elementSize be 1.
+        let mut element_size = 1;
+
+        // Let ctor be %DataView%.
+        let mut ctor = Constructor::DataView;
+
+        // If view has a [[TypedArrayName]] internal slot (i.e., it is not a DataView),
+        if view.has_typed_array_name() {
+            // Set elementSize to the element size specified in the
+            // typed array constructors table for view.[[TypedArrayName]].
+            let view_typw = view.get_array_buffer_view_type();
+            element_size = byte_size(view_typw);
+
+            // Set ctor to the constructor specified in the typed array constructors table for view.[[TypedArrayName]].
+            ctor = Constructor::Name(view_typw);
+        }
+
+        // Let minimumFill be min × elementSize.
+        let minimum_fill = options.min * element_size;
+
+        // Assert: minimumFill ≥ 0 and minimumFill ≤ view.[[ByteLength]].
+        assert!(minimum_fill <= (view.byte_length() as u64));
+
+        // Assert: the remainder after dividing minimumFill by elementSize is 0.
+        assert_eq!(minimum_fill % element_size, 0);
+
+        // Let byteOffset be view.[[ByteOffset]].
+        let byte_offset = view.get_byte_offset();
+
+        // Let byteLength be view.[[ByteLength]].
+        let byte_length = view.byte_length();
+
+        // Let bufferResult be TransferArrayBuffer(view.[[ViewedArrayBuffer]]).
+        let buffer_result = view
+            .get_array_buffer_view_buffer(cx)
+            .transfer_array_buffer(cx);
+
+        // If bufferResult is an abrupt completion,
+        if buffer_result.is_none() {
+            // Perform readIntoRequest’s error steps, given bufferResult.[[Value]].
+            unsafe {
+                rooted!(in(*cx) let mut rval = UndefinedValue());
+                assert!(JS_GetPendingException(*cx, rval.handle_mut()));
+                JS_ClearPendingException(*cx);
+
+                read_into_request.error_steps(rval.handle(), can_gc);
```

Comment:
Perhaps in a follow-up, `transfer_array_buffer` should take a safe mutable handle, and set the error internally.